### PR TITLE
Allow Asset Tx Chaining

### PR DESCRIPTION
**Pull request by blondfrogs**
This PR allows the following to be done with assets before a block is mined. 
``` issue CHAIN 100 , issue CHAIN/A 100, issue CHAIN/B 100 ```
``` issue CHAIN/A/A ,issue CHAIN/A/A/C ```
You can also chain reissues of a asset once it is issued in a block
Block Mined that contains the above issues
``` reissue CHAIN 100 "address", reissue CHAIN 200 "address" ``` two or more reissues are valid to the chain

Enforce all asset transaction OutPoints to contain 0 RVN in them. This will make it so the network wont allow tx's to accidentally spend there assets when spending RVN to an address. 

Updated AvailableCoins function to return only asset tx's or rvn tx's depending on flags passed to function. This fixes the asset outpoints showing up in the Send Dialog Coincontrol. Only RVN tx's show up in the Send Dialog Coincontrol box.

### CODE REVIEW

File: src/assets/assets.cpp [hunk 1]
```diff
@@ -776,10 +835,12 @@ bool CAssetsCache::TrySpendCoin(const COutPoint& out, const CTxOut& txOut)
     if (address != "" && assetName != "" && nAmount > 0) {
         CAssetCacheSpendAsset spend(assetName, address, nAmount);
         if (GetBestAssetAddressAmount(*this, assetName, address)) {
-            assert(mapAssetsAddressAmount[make_pair(assetName, address)] >= nAmount);
-            mapAssetsAddressAmount[make_pair(assetName, address)] -= nAmount;
+            auto pair = make_pair(assetName, address);
+            mapAssetsAddressAmount.at(pair) -= nAmount;
 
-            if (mapAssetsAddressAmount[make_pair(assetName, address)] == 0 &&
+            if (mapAssetsAddressAmount.at(pair) < 0)
```
Comment (cfox) [inline src/assets/assets.cpp]: This should never happen right?

File: src/consensus/tx_verify.cpp [hunk 2]
```diff
@@ -401,22 +415,27 @@ bool Consensus::CheckTxAssets(const CTransaction& tx, CValidationState& state, c
...
                 }
             }
+        }
+    }
 
+    for (const auto& outValue : totalOutputs) {
+        if (!totalInputs.count(outValue.first)) {
+            std::string errorMsg;
+            errorMsg = strprintf("Bad Transaction - Trying to create outpoint for asset that you don't have: %s", outValue.first);
+            return state.DoS(100, false, REJECT_INVALID, "bad-tx-inputs-outputs-mismatch " + errorMsg);
```
Comment (cfox) [inline src/consensus/tx_verify.cpp]: copy/pasted error code

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2137,23 +2146,34 @@ CAmount CWallet::GetAvailableBalance(const CCoinControl* coinControl) const
...
-    AvailableCoinsAll(vCoins, mapAssetCoins, setAssetOutPoint, fWithAssets, fOnlySafe, coinControl, nMinimumAmount, nMaximumAmount, nMinimumSumAmount, nMaximumCount, nMinDepth, nMaxDepth);
+    AvailableCoinsAll(vCoins, mapAssetCoins, false, true, fOnlySafe, coinControl, nMinimumAmount, nMaximumAmount, nMinimumSumAmount, nMaximumCount, nMinDepth, nMaxDepth);
 }
 
+void CWallet::AvailableCoinsWithAssets(std::vector<COutput> &vCoins, std::map<std::string, std::vector<COutput> > &mapAssetCoins,
+                              bool fOnlySafe, const CCoinControl *coinControl, const CAmount &nMinimumAmount,
+                              const CAmount &nMaximumAmount, const CAmount &nMinimumSumAmount,
+                              const uint64_t &nMaximumCount, const int &nMinDepth, const int &nMaxDepth) const
+{
+    if (!AreAssetsDeployed())
```
Comment (cfox) [inline src/wallet/wallet.cpp]: Should this still return RVN coins instead of bailing?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -2684,75 +2717,223 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
...
+            nTotalLower += nTempAmount;
         }
+        else if (!coinLowestLarger || !coinLowestLargerAmount || nTempAmount < coinLowestLargerAmount)
+        {
+            coinLowestLarger = coin;
+            coinLowestLargerAmount = nTempAmount;
+        }
+
+
+//        setCoinsRet.insert(CInputCoin(out.tx, out.i));
```
Comment (cfox) [inline src/wallet/wallet.cpp]: commented section still needed?

File: src/consensus/tx_verify.cpp (hunk 2 above)
Comment (blondfrogs) [reply src/consensus/tx_verify.cpp]: The return statements are the same yes, but they return different messages.

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (blondfrogs) [reply src/wallet/wallet.cpp]: Removed the commented out code

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (blondfrogs) [reply src/wallet/wallet.cpp]: Updated with better code. This is a copy paste mistake

File: src/assets/assets.cpp (hunk 1 above)
Comment (blondfrogs) [reply src/assets/assets.cpp]: Can happen, only with chained transactions. It occurs when blocks are being undone because of a chain split

**Pull request merged**